relay-filter/src/config.rs

```diff
@@ -328,15 +482,215 @@ impl FiltersConfig {
+    /// Index of the next filter in global config to evaluate.
+    global_index: usize,
+    /// Filters that have been evaluated, either yielded or ignored.
+    evaluated: HashSet<&'a String>,
+}
+
+impl<'a> DynamicFiltersConfigIter<'a> {
+    /// Creates an iterator over the filters in [`DynamicFiltersConfig`].
+    pub fn new(config: DynamicFiltersConfig<'a>) -> Self {
+        DynamicFiltersConfigIter {
+            config,
+            project_index: 0,
+            global_index: 0,
+            evaluated: HashSet::new(),
+        }
+    }
+}
+
+impl<'a> DynamicFiltersConfigIter<'a> {
+    /// Returns whether the inbound filters support the maximum version.
+    ///
+    /// Filters are supported if the versions of filters of both project and
+    /// global configs are not greater than the given maximum version.
+    ///
+    /// Filters from the project and global configs are complementary, and in
+    /// isolation, they don't provide enough valuable information to perform the
+    /// filtering. Additionally, new versions may include features not supported
+    /// in the current Relay.
+    fn is_version_supported(&self) -> bool {
+        if self.config.max_version < self.config.project_filters.version {
+            return false;
+        }
+        if let Some(gc) = self.config.global_filters {
+            if self.config.max_version < gc.version {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn next_project_filter(&mut self) -> Option<&'a GenericFilterConfig> {
+        let Some((id, filter)) = self
+            .config
+            .project_filters
+            .filters
+            .get_index(self.project_index)
+        else {
+            return None;
+        };
+        self.project_index += 1;
+
+        // Mark the filter as evaluated first. If the filter is disabled, it
+        // should not be applied from global configs.
+        let newly_evaluated = self.evaluated.insert(id);
+
+        if !filter.is_enabled || !newly_evaluated {
+            return None;
+        }
+
+        if filter.condition.is_none() {
+            let Some(global_filters) = self.config.global_filters else {
+                return None;
+            };
+            let Some(filter) = global_filters.filters.get(id) else {
+                return None;
+            };
+            return Some(filter);
+        }
+
+        Some(filter)
+    }
+
+    fn next_global_filter(&mut self) -> Option<&'a GenericFilterConfig> {
+        let Some(global_filters) = self.config.global_filters else {
+            return None;
+        };
+        let Some((id, filter)) = global_filters.filters.get_index(self.global_index) else {
+            return None;
+        };
+        self.global_index += 1;
+
+        if filter.is_empty() {
+            return None;
+        }
+
+        let newly_added = self.evaluated.insert(id);
+        if !newly_added {
+            return None;
+        }
+
+        Some(filter)
```

Comment:
```suggestion
        self.evaluated.insert(id).then_some(filter)
```

(Maybe I like combinators too much)